### DISCUSSION TITLE
Maximum bit depth for RGB

### DISCUSSION THREAD

**Discussion by SmForker** [ultralytics/ultralytics#22922] [Q&A] [answered]
Does Yolo12 (or other) round up values for the input image  - 48-bit RGB (16 bits per channel) to 24-bit (8 bits per channel)?
If so, are there ways to fix it?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @SmForker, thank you for your interest in Ultralytics and YOLO 🚀  
This is an automated response to help you get started while an Ultralytics engineer reviews your question and follows up soon.

For detailed usage and behavior of YOLO models (including input handling and preprocessing), we recommend checking the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples. Many common questions about image formats, normalization, and data pipelines are addressed there and may help clarify how inputs are processed under the hood.

If this turns out to be a 🐛 Bug Report (for example, if you observe unexpected behavior when using high bit-depth images versus standard 8-bit images), please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so the team can investigate efficiently.

If this is a custom training or data-processing ❓ Question (such as using non-standard image formats, bit depths, or custom dataloaders), please share:
-Your exact workflow (inference vs. training, and how you load images)  
-Relevant code snippets or config  
-Example images (or description of format/bit depth) and any logs or console output  

You can also verify that your setup and data pipeline follow our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/), which include guidance on image preparation and consistency ✅

Join the Ultralytics community where it suits you best:
-For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧  
-For in-depth, longer-form discussions, visit [Discourse](https://community.ultralytics.com/)  
-Or explore and share knowledge on our [Subreddit](https://www.reddit.com/r/Ultralytics/)  

## Upgrade

Before digging deeper, please ensure you are on the latest `ultralytics` package (this helps rule out issues already fixed in recent releases) and that all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) are satisfied in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply, marked as answer]
Short answer: all Ultralytics YOLO models (including YOLO12 and YOLO11) effectively work with 8‑bit‑style inputs, not true 16‑bit per channel.  

Images are converted to float tensors and normalized assuming a 0–255 range, for example in the `ToTensor` transform which divides by `255.0` and outputs values in `[0, 1]` regardless of the original on‑disk bit depth ([ToTensor docs](https://docs.ultralytics.com/reference/data/augment/#totensor)). If you pass a `torch.Tensor` directly, the loader checks and, when it sees values >1, will similarly divide by 255.0 to normalize ([LoadTensor docs](https://docs.ultralytics.com/reference/data/loaders/#loadtensor)).  

So even if you start from a 48‑bit RGB TIFF/PNG, the extra precision is not preserved in the default pipeline; there isn’t a built‑in switch to run the full model at 16‑bit/channel. If you really need to control this, the current workaround is to handle loading and scaling yourself, e.g. read with OpenCV using `IMREAD_UNCHANGED`, convert to `float32`, normalize by `65535.0` to `[0, 1]`, transpose to `BCHW`, and pass that tensor to `model(...)` so the internal normalization doesn’t re‑scale it. Just note that the pretrained weights were trained on 8‑bit imagery, so you typically won’t gain much from 16‑bit inputs.
